Delete memberships when a group is deleted

pfrenssen opened the pull request:
When a group is deleted we also need to delete the memberships. This was originally handled in `og_entity_delete()`. The roles and content was also deleted, but let's do this in separate PRs.

There are two ways of deleting memberships, depending on how large the site is: queued and non-queued. There is also a mechanism of preventing the deletion from taking place, to allow developers to provide their own implementation.

jhedstrom committed:
Split OgDeleteOrphansTest into a separate class.

pfrenssen committed:
Start implementing the membership deletion from the D7 version.

pfrenssen commented:
We need this for our project. I plan on starting on this tomorrow. I would like to get some feedback on the approach.

In Drupal 7 there are three ways of dealing with the deletion of memberships (ref. [og_membership_delete_by_group()](http://www.drupalcontrib.org/api/drupal/contributions!og!og.module/function/og_membership_delete_by_group/7):
- Do not perform a deletion: activated by setting a `skip_og_membership_delete_by_group` property on the group entity object. This is intended to allow developers to provide their own implementation.
- Perform a batch deletion: when the 'use_queue' option is selected in the configuration.
- Perform a straight deletion with `entity_delete_multiple()`: when the queue option is not selected.

For D8 would it be a good idea to use a service and a plugin type for this? We can provide an interface and two implementations: one for deleting by batch process, one without. Developers can provide their own implementation by providing a class of their own.

In the UI instead of a checkbox we can provide a list of all implementations which the administrator can choose from. This would be a very clean and extensible solution, giving good feedback to the site administrator, and not forcing people to hack the entity object to alter the behavior. Other developers can also easily extend our classes to make minor modifications if desired.

RoySegall commented:
I wrote this one in the D7 branch so I know that batch operation sound good but we came to realise that cron or queue worker are the best solution. When will you fire up the batch operation? From a UX perspective i don't this is smart that normal user will see the batch.

Don't think plugin is a good idea but what is was thinking of it for now is do something like the node access rebuild. If you were in the admin panel you could "node permission need to be rebuild" and then you will be redirected to the batch process.

pfrenssen commented:
I thought of making it a plugin because we have multiple implementations of the same thing. That fits really well with the concept of plugins. As it is said in the documentation:

> Plugins are small pieces of functionality that are swappable.
> 
> If you expose a UI through which people can configure or select which implementation they want, you should use the plugin system - even if it is mostly the implementation details only which vary.

I like the idea of using cron. That would hide the process entirely. Still from experience I know that cron is problematic for some sites, sometimes they need to use Poor Man's Cron and then the burden is placed on a random visitor of the website.

I think it all depends on the site in question, sites with little content might want to do it all at once. Sites with a lot of content but only a few admins might prefer a batch operation. Sites with many editors will want a cron job. Other sites might want their own custom behavior.

So from an OO perspective we should make something like a OgDeleteOrphansInterface and a couple of implementations. We also need a manager to interface with the UI. That just sounds like a plugin to me, otherwise we are simply inventing our own plugin handler.

I'll look into the QueueWorkers. I haven't worked with it yet but I see there are a few example implementations in core.

RoySegall commented:
I love plugin. I really do. I glad you said and not me :)

Any way, we need first to plan the base methods. The methods should go like:
Register - notify the deletion worker that there are orphans we need to delete. This should be there because servers could have various of job workers. I don't know if there any but I have a strong feeling they exists.
Query - Something that will do the query. This could be in the base plugin class because developer could connect their entity storage to another DB(rethink DB for example) and they want to override it easily.
Processor - the method will delete the orphan. Might be in the base plugin class and maybe developer would like to send notification to a socket IO server before the deletion.

This might look like an over engineering but if we want to give an easy OO we should split it to small methods. 

Another option is to give a small button in the UI that will start the operation for all the orphans.

pfrenssen commented:
Thanks, that sounds like a great plan. Do you think we should pass in the group entities and let the class derive the orphans from it, or pass in a list of orphan entities directly?

RoySegall commented:
We should give that to query method. But let's do it like in D7 where we register the orphans in the item.

pfrenssen committed:
Create skeleton for deleting orphans.

pfrenssen committed:
Declare the annotation class for the OgDeleteOrphans plugin type.

pfrenssen committed:
Define a basic plugin manager for the OgDeleteOrphans plugin type.

pfrenssen committed:
Allow to select a method to delete orphans in the configuration form.

pfrenssen committed:
Set plugin weights to determine the order they are shown in the administration form.

pfrenssen committed:
Add todos.

pfrenssen committed:
Rename config keys to match the name of the plugin type.

pfrenssen committed:
Squashed #164

pfrenssen committed:
Install Composer dependencies on Drupal 8.1.x.

amitaibu commented on og.module:
```diff
@@ -44,6 +44,24 @@ function og_entity_insert(EntityInterface $entity) {
 /**
+ * Implements hook_entity_delete().
+ */
+function og_entity_delete(EntityInterface $entity) {
```
we actually need to do it now only when a user or a group is deleted

pfrenssen replied:
Indeed! We also have `hook_entity_predelete()` now in Drupal 8, that is more appropriate for this.

amitaibu commented on src/OgDeleteOrphansBase.php:
```diff
@@ -0,0 +1,47 @@
+  /**
+   * {@inheritdoc}
+   */
+  public function configurationForm($form, FormStateInterface $form_state) {
```
Is there a (real world) use case for this?

pfrenssen replied:
I intend to allow the user to configure the number of group content items to delete per batch operation.

amitaibu replied:
I think it might be too much (i.e. give site builers a way to shot their foot). Lets go with some sane number (50), and of someone wants to have control over this - let them extend the plugin?

RoySegall replied:
Defining a new plugin for this one might be intermediate for just change the number of items per batch.

pfrenssen replied:
OK I'll leave it out for now.

pfrenssen replied:
I added it back, the real use case now is the submit button that launches the batch process.

amitaibu replied:
👍

amitaibu commented on src/OgDeleteOrphansInterface.php:
```diff
@@ -0,0 +1,81 @@
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   *   The group entity to register.
+   */
+  public function register(\Drupal\Core\Entity\EntityInterface $entity);
```
Maybe `Register a soon to be deleted group entity, for processing (e.g. deletion of its orphaned members or content).`

amitaibu commented on src/OgDeleteOrphansPluginManager.php:
```diff
@@ -0,0 +1,31 @@
+   * @param \Drupal\Core\Extension\ModuleHandlerInterface $module_handler
+   *   The module handler to invoke the alter hook with.
+   */
+  public function __construct(\Traversable $namespaces, CacheBackendInterface $cache_backend, ModuleHandlerInterface $module_handler) {
```
Do we need the `ModuleHandlerInterface $module_handler` here?

pfrenssen replied:
Yes we need to pass it to the parent constructor.

amitaibu commented on src/Plugin/OgDeleteOrphans/Batch.php:
```diff
@@ -0,0 +1,26 @@
+   * {@inheritdoc}
+   */
+  public function process() {
+    throw new \Exception(__METHOD__ . ' is not implemented.');
```
lets ship it this way, and see the users face when they see it! (evil laugh) :wink:

pfrenssen replied:
We can release the beta1 tomorrow :D

pfrenssen committed:
Rename method, it's obvious that the memberships are deleted by group.

pfrenssen committed:
Delete old Drupal 7 code.

pfrenssen committed:
Implement initial version of hook_entity_predelete().

pfrenssen committed:
Update documentation.

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into delete-memberships

pfrenssen committed:
Delete the old Drupal 7 test.

pfrenssen committed:
Update documentation.

pfrenssen committed:
Disable the test for moving orphaned content to another group. This needs clarification.

pfrenssen committed:
Turn this into a KernelTest.

pfrenssen committed:
Rename test.

pfrenssen committed:
Add a test.

pfrenssen committed:
Small test fixes.

pfrenssen committed:
Create the group and group content in the setup of the test so it can be reused.

pfrenssen committed:
Test that orphaned content is not deleted when the option is disabled.

pfrenssen commented:
The scaffold and test are in place. Going to start the actual implementation tomorrow.

pfrenssen committed:
Implement initial version of Og::deleteMemberships().

pfrenssen commented:
Hmm anyone got experience yet with doing a reverse entity reference lookup? I need to get the list of group content that is referencing the group entity which is being deleted.

This is fun since there can be many group content entity types referencing the same group, so we cannot use a single EntityQuery to do the lookup.

amitaibu commented:
We have something like it in Message D7, I can look for the code when I'll
be in front of a computer
On Mar 16, 2016 4:09 PM, "Pieter Frenssen" notifications@github.com wrote:

> Hmm anyone got experience yet with doing a reverse entity reference
> lookup? I need to get the list of group content that is referencing the
> group entity which is being deleted.
> 
> This is fun since there can be many group content entity types referencing
> the same group, so we cannot use a single EntityQuery to do the lookup.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> https://github.com/amitaibu/og/pull/162#issuecomment-197348672

RoySegall commented:
D7: https://github.com/Gizra/message/blob/7.x-1.x/message.module#L381
D8: https://github.com/Gizra/message/blob/8.x-1.x/message.module#L31

pfrenssen commented:
Thanks! I'll have a peek.

pfrenssen commented:
I've temporarily merged in the ajax fix from #164. It's pretty much impossible to set up group content in the UI without that. I will regularly rebase this branch to keep it clean.

pfrenssen commented:
Also temporarily merged in #163. I need a functioning UI so I can set up test content locally. I will also follow this up with rebases.

pfrenssen committed:
Start implementing a public static method that returns all group content associated with a given group entity.

pfrenssen committed:
Wondering about this. This seems to be the reverse from what is mentioned in the field type descriptions.

pfrenssen committed:
Start on a test for Og::getGroupContent().

pfrenssen committed:
Compile and return the group content.

pfrenssen commented:
I gave up on trying to set up test groups through the UI. I am doing it in a test now but I'm currently confused about how the `target_type` and `target_bundles` are stored. I can find the `target_type` when I load the FieldStorageConfig object, but there is no trace of the bundles (or any of the other `handler_settings`). There is probably a separate config entity for the handler settings. Going to look into that tomorrow.

amitaibu commented:
Thanks for the hard work @pfrenssen !

pfrenssen committed:
Remove the configuration settings again. We don't need them for now.

pfrenssen committed:
Provide a helper method on the base class that returns all content of a given group.

pfrenssen committed:
Start implementing the Simple plugin.

pfrenssen committed:
Mark the Disabled test as skipped until it is ready to be implemented.

pfrenssen commented:
Committed all my work in progress. I have to work on other tasks this week. I can pick this back up next Monday at the earliest.

pfrenssen committed:
Merge branch 'OgComplexWidget-target_type' into delete-memberships

pfrenssen commented:
I merged in the latest changes to #163 since this was my main blocker while working on this ticket. With this fix in I should actually able to create groups and group content freely to test the functionality.

pfrenssen committed:
Fix typo.

pfrenssen committed:
Complete the basic test for Og::getGroupContent().

pfrenssen committed:
Add the column name that we want to retrieve to the condition.

pfrenssen committed:
Rename variable.

pfrenssen committed:
Test that correct results are returned when filtering by entity type.

pfrenssen committed:
Remove the exclusive option. This won't work if group content has multiple group audience fields.

pfrenssen committed:
Update documentation.

pfrenssen committed:
Test retrieval of group content that references multiple groups.

pfrenssen commented:
Ok I got my reverse lookup in place. Now I can start implementing the first plugin. Going to start with the `Simple` plugin.

pfrenssen committed:
Provide static methods that return associated groups for a given group content entity.

pfrenssen committed:
Implement the register() and query() methods on the plugin base class.

pfrenssen committed:
Implement the processor on the Simple plugin.

pfrenssen committed:
Fix test. The reference to the group content is also deleted.

pfrenssen committed:
Temporarily disable the unimplemented plugins so we can see a green test.

pfrenssen committed:
Merge remote-tracking branch 'upstream/8.x-1.x' into delete-memberships

pfrenssen commented:
Major milestone! The first plugin is ready, meaning that the orphaned memberships are now being deleted.

What still needs to be done:
- Implement remaining plugins.
- I have added a few static methods to the `Og` class that retrieve group IDs, group entities and a count of the number of groups that are associated with a given group content entity. These methods still need to be tested.
- Extend test coverage a bit.

@amitaibu, @RoySegall I have two questions:
1. I am starting to doubt my idea to provide a "batch" plugin. I had this idea that we could somehow launch a batch process after deleting a group through the UI but I don't think we can implement this reliably. We would have to launch the batch process in to the submit handler on the entity delete confirmation form, but this comes with potential problems:
   - Different entities might have different forms with their own redirects that might conflict with ours.
   - The batch process wouldn't be launched if groups are deleted programmatically.
   
   The only reliable way would be to provide a button in the administration to clean up the orphans, but it seems a but stupid to ask a manual intervention from an admin every time a group is deleted, even though this is probably a very rare occasion on most sites.
   
   I'm thinking now that having the Simple plugin for small sites, and the Cron plugin for large sites should cover most use cases. What do you think?
2. The `OgDeleteOrphansInterface::query()` method seems unnecessary. It ended up being a one liner which is executed as part of the `::register()` method. Would it be OK to remove this, and only keep `::register()` and `::process()`?

RoySegall commented:
1. I think that the batch plugin is a good idea. it's remind me the rebuild permission batch operation in D7. As I mentioned before this could be triggered via a form.
2. it sounds good to me. if we registering the entities and then later iterating over them we don't need the query.

pfrenssen commented:
I'm now thinking that instead of using the State system to store our list of orphans to process, I should use a Queue instead. There seems to be a drawback in that adding items to the queue does a query for every single item that is added, so in the case that a group is deleted with 1000 group content items, 1000 queries would be executed.

This is probably acceptable behavior. Deleting groups is a rare occasion, and the queries by themselves are very small (ref. `DatabaseQueue::createItem()`). If this becomes a problem in the future this is something that needs to be fixed in core and not by us, for example by providing a `::createItems()` method, or a `::commit()` method to execute the query after adding a multitude of items.

Batch and cron processes in Drupal tie in with Queues, so I'm going ahead and implement it.

amitaibu commented:
> Deleting groups is a rare occasion

Indeed. I think it's fine to use the queue system

pfrenssen committed:
Use a Queue to store the list of orphans to delete.

pfrenssen committed:
Let PluginBase do some of the lifting, no need to reinvent the wheel.

pfrenssen committed:
Remove query() from the interface, turn it into a protected method.

There is not a lot of value in having this as a required method on the interface
since the query is executed in ::register().

pfrenssen committed:
Make deletion of an orphan reusable.

pfrenssen committed:
Fix typo.

pfrenssen committed:
Implement the Cron plugin.

pfrenssen commented:
I'm now using a Queue and have implemented the Cron plugin.

In order to process orphaned group content on cron I implemented a QueueWorker. This now has the side effect that it will always clean up any queued orphans on cron, even if another plugin (like the Batch plugin) is chosen. This has advantages and disadvantages:
- Pro: orphans are always cleaned up after a certain period, even if the admin forgets to manually invoke the batch process.
- Contra: somebody might want to implement a custom plugin where they want full control over the deletion, and this might interfere with it.

What would be best? I can easily change this by providing a unique queue for the cron job.

pfrenssen commented:
Hmm I'm thinking actually that it would be better to only let cron clean up orphans when that option is explicitly chosen. Users might find it very strange that orphans are still cleaned up on cron when this option is not chosen.

pfrenssen committed:
Use a separate queue for the cron job to prevent cron from cleaning up the orphans of other plugins.

pfrenssen commented:
Here's an example implementation for starting a batch execution of a queue: http://cgit.drupalcode.org/queue_ui/tree/src/Form/QueueUIOverviewForm.php?h=8.x-2.x

pfrenssen committed:
Revert "Remove the configuration settings again. We don't need them for now."

This reverts commit caaac09ea35350af07a68f46c7f20d257694b1ce.

pfrenssen committed:
Do not specify validate and submit handlers on the interface. Let plugins manage these themselves.

pfrenssen committed:
Show configuration option for each plugin in the administration screen.

pfrenssen committed:
Implement the batch plugin.

pfrenssen committed:
Remove the original test for Drupal 7.

pfrenssen commented:
Implemented the batch plugin. I have added the button to launch the batch process in the OG_UI settings:

![batch](https://cloud.githubusercontent.com/assets/442924/14288850/1b0cfc56-fb61-11e5-8cfe-9fdac1fa93fe.png)

It's disabled when there are no orphans to delete.

This is now almost complete. I still would like to add some test coverage for the static methods I added to `Og`. That's for tomorrow.

pfrenssen commented on tests/src/Kernel/OgDeleteOrphanedGroupContentTest.php:
```diff
@@ -0,0 +1,212 @@
+   *   child relation which is missing in this test. Here the content just moves
+   *   to whatever random group that is available.
+   */
+  function _testMoveOrphans() {
```
@amitaibu @RoySegall This is an original test for Drupal 7 which was moving group content to another group. The way the test indicates this behavior is expected to work seems wrong to me, can you have a look what the original intention was?

I would like to know if we need to create a followup for this feature, or if I can simply ignore this and remove this test.

amitaibu replied:
yeah, I see that there's something missing there -- telling it where to move. I'll have to refresh my memory..

amitaibu replied:
Ok, found it:

The logic of what should happen, [is here](https://github.com/Gizra/og/blob/719d8f459fb16b6916585e7b445a0119a4057b5a/og.module#L1769-L1775)

And the test makes more sense when you see [this hook](https://github.com/Gizra/og/blob/719d8f459fb16b6916585e7b445a0119a4057b5a/tests/og_test.module#L59-L64) :)

pfrenssen replied:
That is the missing piece indeed! Now, seeing that this is depending on custom code that needs to run in a custom `hook_entity_delete()` implementation, is this something that is planned to be ported to Drupal 8?

amitaibu replied:
I think it should be - it's a nice feature..

pfrenssen replied:
Ok I will adapt the comments and create a follow-up issue.

pfrenssen replied:
Created followup: #178.

pfrenssen commented on tests/src/Kernel/OgDeleteOrphanedGroupContentTest.php:
```diff
@@ -0,0 +1,212 @@
+   */
+  function testDisabled($plugin_id) {
+    $this->markTestSkipped();
+    return;
```
This also still needs to be done.

pfrenssen committed:
Test that no orphaned group content is deleted when the option is disabled.

pfrenssen committed:
Test retrieval of group content with multiple group audience fields.

pfrenssen committed:
Refactor group content registration. This is only useful in one single situation.

pfrenssen committed:
Add references for all todo's.

pfrenssen commented:
Going to work on adding test coverage for `Og::getGroups()`, `Og::getGroupIds()` and `Og::getGroupCount()` now.

pfrenssen committed:
Test the static methods related to retrieving groups associated with given group content.

pfrenssen committed:
Fix bug uncovered by the tests.

pfrenssen commented:
Added the final test coverage. @amitaibu @RoySegall This is ready to be reviewed.

Also pinging my colleagues @sandervd @idimopoulos.

pfrenssen committed:
Move the 'move orphaned nodes' test back to the D7 test so it's retained for later porting.

pfrenssen committed:
Restore the original test info and setup for future reference.

amitaibu commented on src/Annotation/OgDeleteOrphans.php:
```diff
@@ -0,0 +1,49 @@
+   *
+   * @var int optional
+   */
+  public $weight = NULL;
```
do we really need this one? We could just sort by alphabet

pfrenssen replied:
I added it so the Simple plugin which is selected by default would be at the top of the list. It looked a bit strange to have the bottom one selected by default. I can remove it if you want.

amitaibu replied:
Yeah, I think alphabet order is fine + removes a bit of code. Can't imagine many more plugins added there

pfrenssen replied:
Fixed it, it's sorted alphabetically on the English translation though, so for other languages it won't be alphabetical. Is this important?

amitaibu commented on src/Og.php:
```diff
@@ -183,6 +182,166 @@ public static function getEntityGroups(EntityInterface $entity, array $states =
   /**
+   * Returns all group IDs associated with the given group content entity.
+   *
+   * Do not use this to retrieve group IDs associated with a user entity. Use
```
I think that maybe we should take the extra step an throw exception if the entity type is a user - to make it really obvious using it on a user is a bad idea :)

pfrenssen replied:
Good plan!

amitaibu commented on src/Og.php:
```diff
@@ -183,6 +182,166 @@ public static function getEntityGroups(EntityInterface $entity, array $states =
+   * Returns all group IDs associated with the given group content entity.
+   *
+   * Do not use this to retrieve group IDs associated with a user entity. Use
+   * Og::GetEntityGroups() instead.
```
should be `Og::getEntityGroups` - but with the recent changes in OgMembership we can probably call it `Og::getUserEntityGroups` - can be in follow up.

pfrenssen replied:
Fixed the capitalization. There is a follow up to rename it: #169.

amitaibu commented on src/Og.php:
```diff
@@ -183,6 +182,166 @@ public static function getEntityGroups(EntityInterface $entity, array $states =
+      }, $entity->get($field->getName())->getValue());
+
+      // Query the database to get the actual list of groups. The target IDs may
+      // contain groups that no longer exist. Entityreference doesn't clean up
```
Entity reference

pfrenssen replied:
Thanks! Fixed.

amitaibu commented on src/Og.php:
```diff
@@ -183,6 +182,166 @@ public static function getEntityGroups(EntityInterface $entity, array $states =
+  }
+
+  /**
+   * Returns all the group content associated with a given group entity.
```
this should be addressed a bit differently IMO, as it won't scale well. A range must be added here, and sort by ID. so when you need this, you could for example pass the $id to start the query from.

pfrenssen replied:
I had noticed the performance problem during testing and this is no longer returning group content entities, only group content IDs. I have forgotten to update the method name though. This should be called `::getGroupContentIds()` and the documentation should be updated accordingly.

I just tested it with a group containing 10000 group content items and it scales pretty well. The method takes 0.034 seconds on the first run, and 0.013 seconds on the second run (I guess because of the MySql query cache). 

```
$ drush php
Psy Shell v0.7.2 (PHP 7.0.5 — cli) by Justin Hileman
>>> $start = microtime(TRUE); echo count(\Drupal\og\Og::getGroupContent(\Drupal\block_content\Entity\BlockContent::load(5))['node']) . "\n" . (microtime(TRUE) - $start);
10000
0.034804105758667⏎
=> null
>>> $start = microtime(TRUE); echo count(\Drupal\og\Og::getGroupContent(\Drupal\block_content\Entity\BlockContent::load(5))['node']) . "\n" . (microtime(TRUE) - $start);
10000
0.012892007827759⏎
=> null
```

I guess this is acceptable performance :)

amitaibu commented on src/Plugin/OgDeleteOrphans/Batch.php:
```diff
@@ -0,0 +1,75 @@
+   * Batch step definition callback to process one queue item.
+   */
+  public static function step($context) {
+    if (isset($context['interrupted']) && $context['interrupted']) {
```
`if (!empty($context['interrupted'])) {`

pfrenssen replied:
Noob mistake! Fixed :)

amitaibu commented on src/Plugin/OgDeleteOrphans/Cron.php:
```diff
@@ -0,0 +1,36 @@
+  /**
+   * {@inheritdoc}
+   */
+  public function process() {
```
this and the method below confuses me a bit - can you please clarify how it's working + why we use a different queue.

pfrenssen replied:
I reworked it a bit to make it cleaner. The QueueWorker is now calling the process() method, which as an added bonus has deduplicated the code to delete an orphan. I've also updated the documentation to clarify it.

By the way it can be discussed whether or not we use a separate queue for the cron jobs or not. The way it works is that each QueueWorker is attached to a certain Queue. If there are items in the queue, then the queue worker WILL process them on every cron run, if the "Cron" plugin is selected or not.

Initially I was using the same queue, which has the benefit that if you select the Batch plugin and forget to press the button, then the orphans will still be cleaned up on cron over time.

In a next iteration I picked a different queue because I thought that people would find it strange that their orphans would be cleared up on cron if the option was not chosen.

Which solution would you prefer? We can use the same queue if we mention in the UI that orphans will always be cleaned up on cron.

amitaibu commented on src/Plugin/OgDeleteOrphans/Simple.php:
```diff
@@ -0,0 +1,31 @@
+use Drupal\og\OgDeleteOrphansBase;
+
+/**
+ * Performs a simple deletion of orphans.
```
maybe `Performs an on-the-fly deletion of orphans.` ?

pfrenssen replied:
:+1:

amitaibu commented on src/Plugin/OgFields/AudienceField.php:
```diff
@@ -34,6 +34,10 @@ public function getFieldStorageConfigBaseDefinition(array $values = array()) {
       'settings' => [
         'target_type' => $this->getEntityType(),
       ],
+      // @todo Is this correct? The description of OgStandardReferenceItem says:
```
the comment on OgStandardReferenceItem is wrong.

pfrenssen replied:
Ok, removed the comment, the updating of the description can be handled in #177.

amitaibu commented:
I've added a few comments - but in general, it's a beautiful PR, thank you for all this amazing work!

pfrenssen committed:
Address review remarks.

pfrenssen commented:
Addressed the remarks. Didn't test it though, let's await the verdict of mr. Travis.

pfrenssen committed:
Fix fatal error for disrespecting the interface.

pfrenssen committed:
Sort the plugins alphabetically by plugin key rather tan by an arbitrary weight.

pfrenssen committed:
Call the correct plugin.

pfrenssen committed:
Clean up setUp().

pfrenssen commented:
I've also implemented the cleaning up of user memberships, but I have done this in a separate ticket so that this doesn't complicate this any further: #179.

amitaibu commented:
Great stuff, merged!

pfrenssen commented:
Wooohooo!!! Thanks!!

damiankloip commented:
Now go and have a beer!

On 8 April 2016 at 16:08, Pieter Frenssen notifications@github.com wrote:

> Wooohooo!!! Thanks!!
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> https://github.com/amitaibu/og/pull/162#issuecomment-207471079

pfrenssen commented:
I will! :beers:
